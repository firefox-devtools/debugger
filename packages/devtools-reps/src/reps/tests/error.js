/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* global jest */
const { shallow } = require("enzyme");
const {
  REPS,
  getRep
} = require("../rep");

const {
  expectActorAttribute
} = require("./test-helpers");

const { ErrorRep } = REPS;
const { MODE } = require("../constants");
const stubs = require("../stubs/error");

describe("Error - Simple error", () => {
  // Test object = `new Error("Error message")`
  const stub = stubs.get("SimpleError");

  it("correctly selects Error Rep for Error object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for simple error", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
    expectActorAttribute(renderedComponent, stub.actor);
  });

  it("renders with expected text for simple error in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("Error");
  });
});

describe("Error - Multi line stack error", () => {
  /*
   * Test object = `
   *   function errorFoo() {
   *     errorBar();
   *   }
   *   function errorBar() {
   *     console.log(new Error("bar"));
   *   }
   *   errorFoo();`
   */
  const stub = stubs.get("MultilineStackError");

  it("correctly selects the Error Rep for Error object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for Error object", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders expected text for simple error with a multiple line in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("Error");
  });
});

describe("Error - Error without stacktrace", () => {
  const stub = stubs.get("ErrorWithoutStacktrace");

  it("correctly selects the Error Rep for Error object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for Error object", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent.text()).toEqual("Error: \"Error message\"");
  });

  it("renders expected text for simple error without stacktrace in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("Error");
  });
});

describe("Error - Eval error", () => {
  // Test object = `new EvalError("EvalError message")`
  const stub = stubs.get("EvalError");

  it("correctly selects the Error Rep for EvalError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for an EvalError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for an EvalError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("EvalError");
  });
});

describe("Error - Internal error", () => {
  // Test object = `new InternalError("InternalError message")`
  const stub = stubs.get("InternalError");

  it("correctly selects the Error Rep for InternalError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for an InternalError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for an InternalError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("InternalError");
  });
});

describe("Error - Range error", () => {
  // Test object = `new RangeError("RangeError message")`
  const stub = stubs.get("RangeError");

  it("correctly selects the Error Rep for RangeError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for RangeError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for RangeError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("RangeError");
  });
});

describe("Error - Reference error", () => {
  // Test object = `new ReferenceError("ReferenceError message"`
  const stub = stubs.get("ReferenceError");

  it("correctly selects the Error Rep for ReferenceError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for ReferenceError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for ReferenceError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("ReferenceError");
  });
});

describe("Error - Syntax error", () => {
  // Test object = `new SyntaxError("SyntaxError message"`
  const stub = stubs.get("SyntaxError");

  it("correctly selects the Error Rep for SyntaxError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for SyntaxError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for SyntaxError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("SyntaxError");
  });
});

describe("Error - Type error", () => {
  // Test object = `new TypeError("TypeError message"`
  const stub = stubs.get("TypeError");

  it("correctly selects the Error Rep for TypeError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for TypeError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for TypeError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("TypeError");
  });
});

describe("Error - URI error", () => {
  // Test object = `new URIError("URIError message")`
  const stub = stubs.get("URIError");

  it("correctly selects the Error Rep for URIError object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for URIError", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with expected text for URIError in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("URIError");
  });
});

describe("Error - DOMException", () => {
  const stub = stubs.get("DOMException");

  it("correctly selects Error Rep for Error object", () => {
    expect(getRep(stub)).toBe(ErrorRep.rep);
  });

  it("renders with expected text for DOMException", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent.text()).toEqual(
      "DOMException: \"'foo;()bar!' is not a valid selector\""
    );
  });

  it("renders with expected text for DOMException in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent.text()).toEqual("DOMException");
  });
});

describe("Error - base-loader.js", () => {
  const stub = stubs.get("base-loader Error");

  it("renders as expected without mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected in tiny mode", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub,
      mode: MODE.TINY
    }));

    expect(renderedComponent).toMatchSnapshot();
  });
});

describe("Error - longString stacktrace", () => {
  const stub = stubs.get("longString stack Error");

  it("renders as expected", () => {
    const renderedComponent = shallow(ErrorRep.rep({
      object: stub
    }));

    expect(renderedComponent).toMatchSnapshot();
  });
});

describe("Error - stacktrace location click", () => {
  it("Calls onViewSourceInDebugger with the expected arguments", () => {
    const onViewSourceInDebugger = jest.fn();
    const object = stubs.get("base-loader Error");

    const renderedComponent = shallow(ErrorRep.rep({
      object,
      onViewSourceInDebugger
    }));

    const locations = renderedComponent.find(".objectBox-stackTrace-location");
    expect(locations.exists()).toBeTruthy();

    expect(locations.first().prop("title")).toBe("View source in debugger â†’ " +
      "resource://devtools/shared/client/debugger-client.js:856:9");
    locations.first().simulate("click", {
      type: "click",
      stopPropagation: () => {},
    });

    expect(onViewSourceInDebugger.mock.calls.length).toEqual(1);
    let mockCall = onViewSourceInDebugger.mock.calls[0][0];
    expect(mockCall.url).toEqual("resource://devtools/shared/client/debugger-client.js");
    expect(mockCall.line).toEqual(856);
    expect(mockCall.column).toEqual(9);

    expect(locations.last().prop("title")).toBe("View source in debugger â†’ " +
      "resource://devtools/shared/ThreadSafeDevToolsUtils.js:109:14");
    locations.last().simulate("click", {
      type: "click",
      stopPropagation: () => {},
    });

    expect(onViewSourceInDebugger.mock.calls.length).toEqual(2);
    mockCall = onViewSourceInDebugger.mock.calls[1][0];
    expect(mockCall.url).toEqual("resource://devtools/shared/ThreadSafeDevToolsUtils.js");
    expect(mockCall.line).toEqual(109);
    expect(mockCall.column).toEqual(14);
  });

  it("Does not call onViewSourceInDebugger on excluded urls", () => {
    const onViewSourceInDebugger = jest.fn();
    const object = stubs.get("URIError");

    const renderedComponent = shallow(ErrorRep.rep({
      object,
      onViewSourceInDebugger
    }));

    const locations = renderedComponent.find(".objectBox-stackTrace-location");
    expect(locations.exists()).toBeTruthy();
    expect(locations.first().prop("title")).toBe(undefined);

    locations.first().simulate("click", {
      type: "click",
      stopPropagation: () => {},
    });

    expect(onViewSourceInDebugger.mock.calls.length).toEqual(0);
  });

  it("Does not throw when onViewSourceInDebugger props is not provided", () => {
    const object = stubs.get("base-loader Error");

    const renderedComponent = shallow(ErrorRep.rep({
      object,
    }));

    const locations = renderedComponent.find(".objectBox-stackTrace-location");
    expect(locations.exists()).toBeTruthy();
    expect(locations.first().prop("title")).toBe(undefined);

    locations.first().simulate("click", {
      type: "click",
      stopPropagation: () => {},
    });
  });
});
