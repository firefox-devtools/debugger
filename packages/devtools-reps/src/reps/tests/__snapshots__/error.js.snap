// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error - Eval error renders with expected text for an EvalError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1022"
>
  EvalError: "EvalError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:10:13
    </span>
  </span>
</span>
`;

exports[`Error - Internal error renders with expected text for an InternalError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1023"
>
  InternalError: "InternalError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:11:13
    </span>
  </span>
</span>
`;

exports[`Error - Multi line stack error renders with expected text for Error object 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1021"
>
  Error: "bar"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      errorBar
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:6:15
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn1"
    >
      errorFoo
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location1"
    >
      debugger eval code:3:3
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn2"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location2"
    >
      debugger eval code:8:1
    </span>
  </span>
</span>
`;

exports[`Error - Range error renders with expected text for RangeError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1024"
>
  RangeError: "RangeError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:12:13
    </span>
  </span>
</span>
`;

exports[`Error - Reference error renders with expected text for ReferenceError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1025"
>
  ReferenceError: "ReferenceError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:13:13
    </span>
  </span>
</span>
`;

exports[`Error - Simple error renders with expected text for simple error 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1020"
>
  Error: "Error message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:1:13
    </span>
  </span>
</span>
`;

exports[`Error - Syntax error renders with expected text for SyntaxError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1026"
>
  SyntaxError: "SyntaxError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:14:13
    </span>
  </span>
</span>
`;

exports[`Error - Type error renders with expected text for TypeError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1027"
>
  TypeError: "TypeError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:15:13
    </span>
  </span>
</span>
`;

exports[`Error - URI error renders with expected text for URIError 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1028"
>
  URIError: "URIError message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      &lt;anonymous&gt;
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      debugger eval code:16:13
    </span>
  </span>
</span>
`;

exports[`Error - base-loader.js renders as expected in tiny mode 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1020"
>
  Error
</span>
`;

exports[`Error - base-loader.js renders as expected without mode 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn1.child1/obj1020"
>
  Error: "Error message"
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      onPacket
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      resource://devtools/shared/client/debugger-client.js:856:9
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn1"
    >
      send
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location1"
    >
      resource://devtools/shared/transport/transport.js:569:13
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn2"
    >
      makeInfallible
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location2"
    >
      resource://devtools/shared/ThreadSafeDevToolsUtils.js:109:14
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn3"
    >
      makeInfallible
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location3"
    >
      resource://devtools/shared/ThreadSafeDevToolsUtils.js:109:14
    </span>
  </span>
</span>
`;

exports[`Error - longString stacktrace renders as expected 1`] = `
<span
  className="objectBox-stackTrace"
  data-link-actor-id="server1.conn2.child1/obj33"
>
  Error: ""
  

  <span
    className="objectBox-stackTrace-grid"
    key="stack"
  >
    <span
      className="objectBox-stackTrace-fn"
      key="fn0"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location0"
    >
      angular/common/esm5/common.js:2656:27
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn1"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location1"
    >
      angular/core/esm5/core.js:12581:9
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn2"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location2"
    >
      angular/core/esm5/core.js:14109:20
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn3"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location3"
    >
      angular/core/esm5/core.js:14052:16
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn4"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location4"
    >
      angular/core/esm5/core.js:14945:55
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn5"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location5"
    >
      angular/core/esm5/core.js:14886:13
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn6"
    >
      View_MetaTableComponent_6
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location6"
    >
      ng:///AppModule/MetaTableComponent.ngfactory.js:98:5
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn7"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location7"
    >
      angular/core/esm5/core.js:14871:12
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn8"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location8"
    >
      angular/core/esm5/core.js:14018:5
    </span>
    <span
      className="objectBox-stackTrace-fn"
      key="fn9"
    >
      node_modules
    </span>
    <span
      className="objectBox-stackTrace-location"
      key="location9"
    >
      angular/core/esm5/core.js:14369:21
    </span>
  </span>
</span>
`;
